src/Microsoft.VisualStudio.Threading/JoinableTask.cs

```diff
@@ -1056,77 +1024,31 @@ private bool TryDequeueSelfOrDependencies(bool onMainThread, HashSet<JoinableTas
         }
 
         /// <summary>
-        /// Adds a <see cref="JoinableTask"/> instance as one that is relevant to the async operation.
+        /// Adds the specified flags to the <see cref="state"/> field.
         /// </summary>
-        /// <param name="joinChild">The <see cref="JoinableTask"/> to join as a child.</param>
-        internal JoinRelease AddDependency(JoinableTask joinChild)
+        private void AddStateFlags(JoinableTaskFlags flags)
         {
-            Requires.NotNull(joinChild, nameof(joinChild));
-            if (this == joinChild)
-            {
-                // Joining oneself would be pointless.
-                return default(JoinRelease);
-            }
-
-            using (this.Factory.Context.NoMessagePumpSynchronizationContext.Apply())
+            // Try to avoid taking a lock if the flags are already set appropriately.
+            if ((this.state & flags) != flags)
             {
```

Comment:
I think you should use Volatile.Read here (or add the `volatile` modifier to the field) to do this safely.
[https://stackoverflow.com/questions/1964731/the-need-for-volatile-modifier-in-double-checked-locking-in-net](https://stackoverflow.com/questions/1964731/the-need-for-volatile-modifier-in-double-checked-locking-in-net)

Reply:
fixed.